### PR TITLE
refactor: Mejora generación de enlaces de pago

### PR DESCRIPTION
**Pull request by TextC0de**
Este PR introduce una refactorización en el sistema de generación de enlaces de pago y la gestión de productos, mejorando la robustez y mantenibilidad del código.

El cambio está motivado por la adición del sistema de Addons en próximos PRs, lo cual requerirá más cambios en la generación de enlaces de pago, este PR facilitará la extensión de dicha funcionalidad.

## Cambios principales
- Separa la lógica de generación de enlaces de pago en funciones dedicadas
- Mejora el tipado con `NonNullableFields` y tipos más estrictos
- Optimiza la creación de productos en Stripe y MercadoPago
- Mejora el manejo de errores para divisas no soportadas y precios faltantes
- Agrega pruebas para los distintos escenarios de generación de enlaces de pago

## Detalles técnicos
- Reorganiza la función `createPaymentIntent` en módulos más pequeños y mantenibles
- Implementa validaciones más robustas para los productos antes de la creación
- Mejora el manejo de productos gratuitos y mixtos (gratuitos/pagados)
- Añade tipos específicos para los estados de las órdenes de compra

## Pruebas
- Se agregan tests para `handlePaymentLinkGeneration`
- Se cubren escenarios de:
  - Generación de enlaces para órdenes válidas
  - Manejo de tickets gratuitos
  - Pagos con MercadoPago y Stripe
  - Casos de error y edge cases
  - Mezcla de tickets gratuitos y pagados

## Breaking Changes
- Se renombra `createPaymentIntent` a `handlePaymentLinkGeneration`
- Se modifican las firmas de las funciones relacionadas con la creación de productos

### CODE REVIEW

File: src/schema/ticket/helpers.ts
Comment (TextC0de) [inline src/schema/ticket/helpers.ts:1]: Ahora se pueden pasar varios tickets y se aprovecha mejora la concurrencia a través de Promise.all.
Por otro lado, se devuelven los tickets actualizados.

File: src/schema/purchaseOrder/actions.tsx [hunk 1]
```diff
@@ -337,10 +381,6 @@ export const createPaymentIntent = async ({
   currencyId: string;
   logger: Logger;
 }) => {
-  if (!USER) {
```
Comment (TextC0de) [inline src/schema/purchaseOrder/actions.tsx:340]: El USER está tipado como not null, no hace falta este check

File: src/schema/purchaseOrder/actions.tsx [hunk 2]
```diff
@@ -374,22 +415,13 @@ export const createPaymentIntent = async ({
...
-  const query = await fetchPurchaseOrderInformation(purchaseOrderId, DB);
-
-  let totalAmount = 0;
-  let allTicketsAreFree = true;
-
-  for (const ticket of query) {
-    if (!ticket.ticketTemplate.isFree) {
-      allTicketsAreFree = false;
-    }
+  const purchaseOrderInfo = await fetchPurchaseOrderInformation(
```
Comment (TextC0de) [inline src/schema/purchaseOrder/actions.tsx:418]: El nombre "query" para esta constante no era muy transparente sobre su contenido

File: src/schema/ticket/helpers.ts [hunk 3]
```diff
@@ -2,51 +2,102 @@ import { eq } from "drizzle-orm";
 import Stripe from "stripe";
 
 import { TRANSACTION_HANDLER } from "~/datasources/db";
-import { selectTicketSchema, ticketsSchema } from "~/datasources/db/schema";
+import { ticketsSchema } from "~/datasources/db/schema";
 import { createOrUpdateStripeProductAndPrice } from "~/datasources/stripe";
+import { applicationError, ServiceErrors } from "~/errors";
 import { Logger } from "~/logging";
 
 export const ensureProductsAreCreated = async ({
```
Comment (TextC0de) [inline src/schema/ticket/helpers.ts:10]: En vez de soportar solo un ticket, permite mayor concurrencia pudiendo recibir varios productos a la vez

File: src/datasources/db/index.ts [hunk 4]
```diff
@@ -26,7 +26,11 @@ export const getDb = ({
   neonUrl: string;
   logger: Logger;
 }) => {
-  const client = postgres(neonUrl);
+  const client = postgres(neonUrl, {
+    onnotice: () => {
+      // DISCARD NOTICES LIKE identifier "x" will be truncated to "y"
```
Comment (joseglego) [inline src/datasources/db/index.ts:31]: Por qué queremos hacer esto? 

Quizás ayuda en Staging / Prod para `debuggear`, entonces quizás es algo que podemos hacer basados en ambiente si genera mucho ruido en local?

File: src/datasources/stripe/index.ts [hunk 5]
```diff
@@ -53,17 +53,19 @@ const getUnitAmount = (unitAmount: number) => {
 
 const STRIPE_RESOURCE_MISSING_ERROR = "resource_missing";
 
+export type CreateOrUpdateStripeProductItem = {
```
Comment (joseglego) [inline src/datasources/stripe/index.ts:56]: Si es un `StripeProductItem` el `stripeId` me parece redundante, alguna razón en especial para cambiarlo?

File: src/schema/purchaseOrder/actions.tsx [hunk 6]
```diff
@@ -239,9 +236,26 @@ const createMercadoPagoPaymentIntent = async ({
...
+
   return {
-    ...intent,
-    totalPrice,
+    paymentPlatform: "mercadopago",
+    paymentPlatformPaymentLink: preference.init_point,
+    paymentPlatformReferenceID: preference.id,
+    paymentPlatformStatus: "none",
+    paymentPlatformExpirationDate: new Date(expirationDate),
+    totalPrice: totalPrice.toString(),
```
Comment (joseglego) [inline src/schema/purchaseOrder/actions.tsx:258]: totalPrice no debería ser numérico?

File: src/schema/purchaseOrder/actions.tsx [hunk 7]
```diff
@@ -25,9 +25,9 @@ import {
 } from "~/datasources/stripe";
 import { applicationError, ServiceErrors } from "~/errors";
 import { Logger } from "~/logging";
-import { ensureProductsAreCreated } from "~/schema/ticket/helpers";
 import { Context } from "~/types";
 
+import { ensureProductsAreCreated } from "../ticket/helpers";
```
Comment (joseglego) [inline src/schema/purchaseOrder/actions.tsx:30]: nit pick , pero es mejor el `~/schema/ticket/helpers` que el `..` (por contexto) - aplica también para `../userTicketsTransfers/helpers` en la línea siguiente

Artículos de referencia:
- https://medium.com/@phongtlam/stop-using-relative-paths-use-import-aliases-for-a-typescript-web-application-45ea8984fd8 
- https://dev.to/naucode/path-aliases-in-typescript-and-why-you-should-use-them-2odf

File: src/schema/purchaseOrder/actions.tsx (hunk 1 above)
Comment (joseglego) [reply src/schema/purchaseOrder/actions.tsx:340]: Estás seguro? Confío en ti!

File: src/schema/purchaseOrder/actions.tsx [hunk 8]
```diff
@@ -352,18 +392,19 @@ export const createPaymentIntent = async ({
...
-    throw new GraphQLError("No authorizado");
+  if (purchaseOrder.userId !== USER.id) {
+    throw new GraphQLError("No autorizado");
   }
 
-  if (!currencyCode) {
-    throw new GraphQLError("No encontramos un currency con ese ID");
+  if (!currency) {
+    throw new GraphQLError(
+      `No encontramos un currency con el id ${currencyId}`,
```
Comment (joseglego) [inline src/schema/purchaseOrder/actions.tsx:406]: 👍

File: src/schema/purchaseOrder/actions.tsx [hunk 9]
```diff
@@ -557,6 +634,36 @@ export const createPaymentIntent = async ({
   };
 };
 
+function calculateTotalAmount(
```
Comment (joseglego) [inline src/schema/purchaseOrder/actions.tsx:637]: 👍

**Review by joseglego — APPROVED** [review]
LGTM

File: src/schema/purchaseOrder/actions.tsx [hunk 10]
```diff
@@ -352,18 +392,19 @@ export const createPaymentIntent = async ({
...
-  const currencyCode = currency?.currency;
 
   if (!purchaseOrder) {
     throw new GraphQLError("Orden de compra no encontrada");
   }
 
-  if (purchaseOrder?.userId !== USER.id) {
-    throw new GraphQLError("No authorizado");
+  if (purchaseOrder.userId !== USER.id) {
+    throw new GraphQLError("No autorizado");
```
[the remark this answers is not in the document]
Comment (TextC0de) [reply src/schema/purchaseOrder/actions.tsx:401]: Lo anoto como deuda técnica post 9punto5

File: src/schema/purchaseOrder/actions.tsx (hunk 1 above)
Comment (TextC0de) [reply src/schema/purchaseOrder/actions.tsx:340]: Sí, seguro. TypeScript garantiza que USER no puede ser null/undefined aquí.

Si en algún punto intentáramos llamar a esta función sin USER o con USER null/undefined, TypeScript nos daría un error de compilación.

Mantener este check en runtime creo que sería redundante y va contra el principio de confianza en el system de tipos.
